File: packages/ui-flutter/lib/src/toggle_group.dart
import 'package:flutter/widgets.dart';

import 'toggle.dart';

/// Mirrors `packages/ui/src/components/toggle-group.tsx`. `variant` /
/// `size` are shared from the group to each item through an
/// `InheritedWidget` (the React Context analogue). `KinetixToggleGroupItem`
/// is a thin pass-through to [KinetixToggle] with those injected. Single-
/// vs multi-select is the caller's to enforce via the per-item callbacks.
class _ToggleGroupScope extends InheritedWidget {
  const _ToggleGroupScope({
    required this.variant,
    required this.size,
    required super.child,
  });

  final KinetixToggleVariant variant;
  final KinetixToggleSize size;

  static _ToggleGroupScope? maybeOf(BuildContext context) =>
      context.dependOnInheritedWidgetOfExactType<_ToggleGroupScope>();

  @override
  bool updateShouldNotify(_ToggleGroupScope oldWidget) =>
      oldWidget.variant != variant || oldWidget.size != size;
}

class KinetixToggleGroup extends StatelessWidget {
  const KinetixToggleGroup({
    super.key,
    required this.children,
    this.variant = KinetixToggleVariant.standard,
    this.size = KinetixToggleSize.md,
  });

  final List<Widget> children;
  final KinetixToggleVariant variant;
  final KinetixToggleSize size;

  @override
  Widget build(BuildContext context) {
    return _ToggleGroupScope(
      variant: variant,
      size: size,
      child: Row(
        mainAxisSize: MainAxisSize.min,
        children: [
          for (var i = 0; i < children.length; i++) ...[
            if (i > 0) const SizedBox(width: 4),
            children[i],
          ],
        ],
      ),
    );
  }
}

class KinetixToggleGroupItem extends StatelessWidget {
  const KinetixToggleGroupItem({
    super.key,
    required this.pressed,
    required this.onChanged,
    required this.child,
  });

  final bool pressed;
  final ValueChanged<bool>? onChanged;
  final Widget child;

  @override
  Widget build(BuildContext context) {
    final scope = _ToggleGroupScope.maybeOf(context);
    return KinetixToggle(
      pressed: pressed,
      onChanged: onChanged,
      variant: scope?.variant ?? KinetixToggleVariant.standard,
      size: scope?.size ?? KinetixToggleSize.md,
      child: child,
    );
  }
}

File: packages/ui-swiftui/Sources/KinetixUI/ToggleGroup.swift
//
// ToggleGroup.swift — KinetixToggleGroup / KinetixToggleGroupItem.
//
// Mirrors packages/ui/src/components/toggle-group.tsx. The React version
// shares `variant` / `size` from the group to each item through React
// Context; custom EnvironmentKeys play the same role here — items don't
// repeat those. KinetixToggleGroupItem is a thin pass-through to
// KinetixToggle with the shared values injected (same call as the Compose
// port). Single- vs multi-select is the caller's to enforce via the
// per-item bindings.
//

import SwiftUI

private struct KinetixToggleGroupVariantKey: EnvironmentKey {
    static let defaultValue = KinetixToggleVariant.default
}

private struct KinetixToggleGroupSizeKey: EnvironmentKey {
    static let defaultValue = KinetixToggleSize.default
}

extension EnvironmentValues {
    var kinetixToggleGroupVariant: KinetixToggleVariant {
        get { self[KinetixToggleGroupVariantKey.self] }
        set { self[KinetixToggleGroupVariantKey.self] = newValue }
    }
    var kinetixToggleGroupSize: KinetixToggleSize {
        get { self[KinetixToggleGroupSizeKey.self] }
        set { self[KinetixToggleGroupSizeKey.self] = newValue }
    }
}

public struct KinetixToggleGroup<Content: View>: View {
    private let variant: KinetixToggleVariant
    private let size: KinetixToggleSize
    private let content: Content

    public init(
        variant: KinetixToggleVariant = .default,
        size: KinetixToggleSize = .default,
        @ViewBuilder content: () -> Content
    ) {
        self.variant = variant
        self.size = size
        self.content = content()
    }

    public var body: some View {
        HStack(spacing: 4) { content } // gap-1
            .environment(\.kinetixToggleGroupVariant, variant)
            .environment(\.kinetixToggleGroupSize, size)
    }
}

public struct KinetixToggleGroupItem<Label: View>: View {
    @Environment(\.kinetixToggleGroupVariant) private var variant
    @Environment(\.kinetixToggleGroupSize) private var size

    @Binding private var isOn: Bool
    private let label: Label

    public init(isOn: Binding<Bool>, @ViewBuilder label: () -> Label) {
        self._isOn = isOn
        self.label = label()
    }

    public var body: some View {
        KinetixToggle(isOn: $isOn, variant: variant, size: size) { label }
    }
}
